ext/ScreenShot/ScreenShot.h

```diff
@@ -5,27 +5,151 @@
+					imgViewInfo.viewType = asset::IImageView<asset::ICPUImage>::ET_2D;
+					imgViewInfo.flags = static_cast<asset::ICPUImageView::E_CREATE_FLAGS>(0u);
+					imgViewInfo.subresourceRange.baseArrayLayer = 0u;
+					imgViewInfo.subresourceRange.baseMipLevel = 0u;
+					imgViewInfo.subresourceRange.layerCount = imgInfo.arrayLayers;
+					imgViewInfo.subresourceRange.levelCount = imgInfo.mipLevels;
+
+					auto imageView = asset::ICPUImageView::create(std::move(imgViewInfo));
+					auto gpuImageView = driver->getGPUObjectsFromAssets(&imageView.get(), &imageView.get() + 1)->front();
+
+					return std::move(gpuImageView);
+				};
+
+				auto gpuImageViewDepthBuffer = createAttachement(false);
+				auto gpuImageViewColorBuffer = createAttachement(true);
+
+				auto frameBuffer = driver->addFrameBuffer();
+				frameBuffer->attach(video::EFAP_DEPTH_ATTACHMENT, std::move(gpuImageViewDepthBuffer));
+				frameBuffer->attach(video::EFAP_COLOR_ATTACHMENT0, std::move(gpuImageViewColorBuffer));
+
+				return frameBuffer;
+			};
+
+			/*
+				Create a ScreenShot with gpu image usage.
+			*/
+
+			bool createScreenShoot(core::smart_refctd_ptr<IrrlichtDevice> device, core::smart_refctd_ptr<video::IGPUImage> gpuImage, const std::string& outFileName)
+			{
+				auto driver = device->getVideoDriver();
+				auto assetManager = device->getAssetManager();
+
+				auto fetchedImageParams = gpuImage->getCreationParameters();
+				auto image = asset::ICPUImage::create(std::move(fetchedImageParams));
+
+				auto trueBufferRowLength = asset::IImageAssetHandlerBase::calcPitchInBlocks(fetchedImageParams.extent.width, asset::getTexelOrBlockBytesize(fetchedImageParams.format));
+
+				auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::ICPUImage::SBufferCopy>>(1u);
+				asset::ICPUImage::SBufferCopy& region = regions->front();
+
+				region.imageSubresource.mipLevel = 0u;
+				region.imageSubresource.baseArrayLayer = 0u;
+				region.imageSubresource.layerCount = 1u;
+				region.bufferOffset = 0u;
+				region.bufferRowLength = trueBufferRowLength;
+				region.bufferImageHeight = 0u;
+				region.imageOffset = { 0u, 0u, 0u };
+				region.imageExtent = image->getCreationParameters().extent;
+
+				auto destinationBuffer = core::smart_refctd_ptr<video::IGPUBuffer>(driver->createDownStreamingGPUBufferOnDedMem(image->getImageDataSizeInBytes()));
+				destinationBuffer->getBoundMemory()->mapMemoryRange(video::IDriverMemoryAllocation::EMCAF_READ, { 0u, destinationBuffer->getSize() });
+
+				driver->copyImageToBuffer(gpuImage.get(), destinationBuffer.get(), 1, regions->begin());
+
+				auto texelBuffer = core::make_smart_refctd_ptr<asset::ICPUBuffer>(image->getImageDataSizeInBytes());
+				auto rawData = reinterpret_cast<uint8_t*>(destinationBuffer->getBoundMemory()->getMappedPointer());
+
+				memcpy(texelBuffer->getPointer(), rawData, image->getImageDataSizeInBytes());
```

Comment:
First, you should be using https://github.com/Crisspl/IrrlichtBAW/pull/32/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h#diff-442e1eb8342a1cee97af9ca34dc05f9cR139 to reduce code duplication

Overall you should be using method outlined here https://github.com/Crisspl/IrrlichtBAW/pull/32/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h#r406730018

But what you did wrong here is that you don't take into account that CPU and GPU execute asynchronously, so before the CPU does anything to mapped memory that the GPU has been writing to, you need to wait on a fence (which has been placed after the operation that would write to the mapped memory) and then if the mapped memory needs to be invalidated before CPU reads it (to invalidate the CPU cache) you should invalidate it - there's a buffer flag you can check for that (Denoiser sample does that).

Reply:
not sure if it's done correctly now, but changed it